src/utils/api/hooks.ts

```diff
@@ -0,0 +1,62 @@
+'use client'
+
+import { useCallback, useEffect, useRef, useState } from 'react'
+
+import { notify } from '..'
+
+export const useApi = <T extends Promise<any>>(
+  fetcher: () => '' | T | false | null | undefined,
+  options: {
+    autofetch?: boolean
+    onBeforeFetch?: () => Promise<void> | void
+    onError?: () => void
+    onResponse?: (res: Awaited<T>) => Promise<void> | void
+    onSuccess?: () => void
+  } = {}
+) => {
+  const [isLoading, setLoading] = useState(false)
+  const [response, setResponse] = useState(null)
+
+  const optionsRef = useRef(options)
+  optionsRef.current = options
+
+  const fetcherRef = useRef(fetcher)
+  fetcherRef.current = fetcher
+
+  const refetch = useCallback(async () => {
+    setLoading(true)
+    await optionsRef.current.onBeforeFetch?.()
+    const res = await fetcherRef.current()
+    if (!res) {
+      setLoading(false)
+      return
+    }
+    await optionsRef.current.onResponse?.(res)
+    if (res.error) {
+      if (optionsRef.current.onError) {
+        await optionsRef.current.onError?.()
+      }
```

Comment:
Will it work like as it's supposed to if you shorten checks like that?
```suggestion
    if (res.error && optionsRef.current.onError) {
        await optionsRef.current.onError?.()
```

Reply:
There will be additional logic soon, so let's keep it as nested if.